### PR TITLE
Add TG support to llama3 family

### PR DESCRIPTION
**Pull request by sraizada-tt**
### Problem description
Llama3 common codebase did not support TG

### What's changed
Added support for TG to Llama 1B, 3B, 8B, 11B text, 70B 

Notes on TG Pipelines: 
- The TG VMs seem to be a bit flimsy.  Falcon7B fails every now and then (even before this commit), and Llama3 might sometimes fail. This cannot be reproduced locally, and I suspect it might be an issue with the host VM.

To alleviate a bit the VM load, we have done the following:
- On frequent pipeline, run `test_llama_model` quick (1L) instead of full model 
- On demo pipeline, run llama3 1B / 8B / 70B. We do not run 3B/11B for now.
- Created issue: https://github.com/tenstorrent/tt-metal/issues/15961

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/12257947523
- [x] T3K pipelines: 
frequent https://github.com/tenstorrent/tt-metal/actions/runs/12259384935
demo https://github.com/tenstorrent/tt-metal/actions/runs/12260055510
unit https://github.com/tenstorrent/tt-metal/actions/runs/12259280565
- [x] Single card pipelines: https://github.com/tenstorrent/tt-metal/actions/runs/12257978488
demo: https://github.com/tenstorrent/tt-metal/actions/runs/12259481145
- [x] TG pipelines:
  - [x] frequent: https://github.com/tenstorrent/tt-metal/actions/runs/12284672219/job/34281619143
  - [x] demo: https://github.com/tenstorrent/tt-metal/actions/runs/12293137019
  - [x] unit: https://github.com/tenstorrent/tt-metal/actions/runs/12282426340

### CODE REVIEW

File: models/common/rmsnorm.py
Comment (mtairum) [inline models/common/rmsnorm.py:52]: Delete?

File: models/common/rmsnorm.py
Comment (mtairum) [inline models/common/rmsnorm.py:94]: Revert?

File: models/demos/llama3/lt
Comment (mtairum) [inline models/demos/llama3/lt:299]: I don't think we need to differentiate if TG here. If you're running tests on an unsupported machine, the tests should not be called. 

@yieldthought  Thoughts?

File: models/demos/llama3/lt
Comment (mtairum) [inline models/demos/llama3/lt:413]: same as above.

File: models/demos/llama3/lt
Comment (mtairum) [inline models/demos/llama3/lt:80]: the hostname is used to get the actual device. Shouldn't we just probe tt-smi instead?

For example `tt-smi -ls` will get you a list of devices:
```
                All available boards on host:                
┏━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ PCI Dev ID ┃ Board Type ┃ Device Series ┃ Board Number    ┃
┡━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ 0          │ Wormhole   │ n300 L        │ 10001451172317c │
│ 1          │ Wormhole   │ n300 L        │ 100014511723132 │
│ 2          │ Wormhole   │ n300 L        │ 100014511723137 │
│ 3          │ Wormhole   │ n300 L        │ 10001421170300e │
│ N/A        │ Wormhole   │ n300 R        │ 10001451172317c │
│ N/A        │ Wormhole   │ n300 R        │ 100014511723132 │
│ N/A        │ Wormhole   │ n300 R        │ 100014511723137 │
│ N/A        │ Wormhole   │ n300 R        │ 10001421170300e │
└────────────┴────────────┴───────────────┴─────────────────┘
                  Boards that can be reset:                  
┏━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ PCI Dev ID ┃ Board Type ┃ Device Series ┃ Board Number    ┃
┡━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ 0          │ Wormhole   │ n300 L        │ 10001451172317c │
│ 1          │ Wormhole   │ n300 L        │ 100014511723132 │
│ 2          │ Wormhole   │ n300 L        │ 100014511723137 │
│ 3          │ Wormhole   │ n300 L        │ 10001421170300e │
└────────────┴────────────┴───────────────┴─────────────────┘
```

Comment (mtairum) [reply models/demos/llama3/lt:80]: And since we use this function to get the actual device name, this doesn't work if you run `lt` on N150/N300 (I know it's uncommon).

**Review by mtairum — CHANGES_REQUESTED** [review]

File: models/demos/llama3/lt
Comment (sraizada-tt) [reply models/demos/llama3/lt:59]: @johanna-rock-tt

File: models/demos/llama3/lt
Comment (sraizada-tt) [reply models/demos/llama3/lt:80]: @johanna-rock-tt

Comment (johanna-rock-tt) [reply models/demos/llama3/lt:80]: Yeah sounds good. I wasn't aware you can get the actual hostname from tt-smi and thought we only run it on t3k and tg, but with this we could just get the hostname and then show whatever fake devices are supported per actual device.

File: models/demos/llama3/lt
Comment (johanna-rock-tt) [reply models/demos/llama3/lt:59]: For TG I got regular hangs at the very start, so the first test just hangs on init devices then. Resetting at start fixes that.

Comment (johanna-rock-tt) [reply models/demos/llama3/lt:59]: There was actually also an issue with cancelling a job where I believe we didn't reset properly and then the next job hangs in init. But I removed this after fixing the code after rebasing because a lot of the code had changed and I didn't run any extended tests after that. Not sure if that's only an issue for TG or in the general case.

Comment (mtairum) [reply models/demos/llama3/lt:59]: I see. We can keep this then 👍 
It's a minor inconvenience for extra reliability.

Do you know if these hangs, assuming they were caused by a device in a bad state, would be visible on tt-smi? If so we can probably modify the code to check tt-smi first and if the device is in bad state, reset at start. 

No strong feeling for this method though, more curious than anything.

Comment (johanna-rock-tt) [reply models/demos/llama3/lt:59]: I don't know if we see it in tt-smi, we could check and change it, yeah. The hang was definitely because of a bad device state since it hung in the init devices already (no model code ru
n at that point).

File: models/demos/llama3/tt/distributed_norm.py
Comment (mtairum) [inline models/demos/llama3/tt/distributed_norm.py:16]: Add a comment explaining these hardcoded values

File: models/demos/llama3/tt/llama_attention.py
Comment (kpaigwar) [inline models/demos/llama3/tt/llama_attention.py:1]: Looks like when running test_model.py for TG with batch_size=32
current_pos tensor/page_table not getting adjusted to batch_per_device (8 in case of TG)

File: models/demos/llama3/tt/llama_rope.py
Comment (kpaigwar) [inline models/demos/llama3/tt/llama_rope.py:174]: Actually we don't need to slice for TG which has separate batch_per_device
We can just initialize `TtLlamaRotarySetup` with `batch_size = batch_per_device`. This setup is only used inside attention.
This is how I have done to integrate fused ROPE where I just 2x batch_size to initialize for Q and K together.

**Comment by mtairum** [comment]
Updated the TG pipelines. 
All passing for Llama3. There are some ND hangs every now and again but this is not related to llama3, we see this for falcon7B even before this commit.

**Review by mtairum — APPROVED** [review]

**Comment by mtairum** [comment]
Review fully done. Just waiting for final approval from infra and then we can merge 🔥

File: .github/workflows/tg-demo-tests-impl.yaml
Comment (tt-rkim) [inline .github/workflows/tg-demo-tests-impl.yaml:17]: Please delete

File: .github/workflows/tg-demo-tests-impl.yaml
Comment (tt-rkim) [inline .github/workflows/tg-demo-tests-impl.yaml:44]: Can we try running in a docker container like `.github/workflows/fast-dispatch-build-and-unit-tests.yaml` so we can get started on this earlier?

File: .github/workflows/tg-frequent-tests-impl.yaml
Comment (tt-rkim) [inline .github/workflows/tg-frequent-tests-impl.yaml:46]: Can we also try running this in docker like previous comment?

**Review by tt-rkim — APPROVED** [review]

File: .github/workflows/tg-demo-tests-impl.yaml
Comment (mtairum) [reply .github/workflows/tg-demo-tests-impl.yaml:43]: I couldn't get it to work.

[This Pipeline](https://github.com/tenstorrent/tt-metal/actions/runs/12314480754) was the longest I got, but everything crashes due to brisc build failed
This was the [commit used](https://github.com/tenstorrent/tt-metal/commit/02ae4b99125bd5b3ed6ce26e177de264d344bce9).

Since the TG pipelines were not using docker previously, we've pushed the PR like that. The Llama3 TG support is higher priority and has been delayed for over a week.

We can create a new smaller issue to tackle the proper usage of docker instances for TG pipelines in the future.

**Pull request merged**
